leopard.nim

```diff
@@ -0,0 +1,329 @@
+       (code.data < MinSymbols) or (code.parity < MinSymbols) or
+       (code.codeword > MaxTotalSymbols))
+
+# alloc/freeAligned and helpers adapted from mratsim/weave:
+# https://github.com/mratsim/weave/blob/master/weave/memory/allocs.nim
+
+func isPowerOfTwo(n: int): bool {.inline.} =
+  (n and (n - 1)) == 0
+
+func roundNextMultipleOf(x: Natural, n: Natural): int {.inline.} =
+  (x + n - 1) and not (n - 1)
+
+when defined(windows):
+  proc aligned_alloc_windows(size, alignment: csize_t): pointer
+    {.header: "<malloc.h>", importc: "_aligned_malloc", sideeffect.}
+
+  proc aligned_free_windows(p: pointer)
+    {.header: "<malloc.h>", importc: "_aligned_free", sideeffect.}
+
+  proc freeAligned*(p: pointer) =
+    if not p.isNil:
+      aligned_free_windows(p)
+
+elif defined(osx):
+  proc posix_memalign(mem: var pointer, alignment, size: csize_t)
+    {.header: "<stdlib.h>", importc, sideeffect.}
+
+  proc aligned_alloc(alignment, size: csize_t): pointer {.inline.} =
+    posix_memalign(result, alignment, size)
+
+else:
+  proc aligned_alloc(alignment, size: csize_t): pointer
+    {.header: "<stdlib.h>", importc, sideeffect.}
+
+when not defined(windows):
+  proc c_free(p: pointer) {.header: "<stdlib.h>", importc: "free".}
+
+  proc freeAligned*(p: pointer) {.inline.} =
+    if not p.isNil:
+      c_free(p)
+
+proc allocAligned*(size: int, alignment: static Natural): pointer {.inline.} =
+  ## aligned_alloc requires allocating in multiple of the alignment.
+  static:
+    assert alignment.isPowerOfTwo()
+
+  let
+    requiredMem = size.roundNextMultipleOf(alignment)
+
+  when defined(windows):
+    aligned_alloc_windows(csize_t requiredMem, csize_t alignment)
+  else:
+    aligned_alloc(csize_t alignment, csize_t requiredMem)
+
+proc leoInit*() =
+  if wrapper.leoInit() != 0:
+    raise (ref LeopardDefect)(msg: "Leopard-RS failed to initialize")
+
+proc encode*(code: ReedSolomonCode, data: Data):
+    Result[ParityData, LeopardError] =
+  if not code.isValid:
+    return err LeopardError(code: LeopardBadCode, msg: LeopardBadCodeMsg)
+
+  var
+    data = data
+
+  let
+    symbolBytes = data[0].len
+
+  if data.len < code.data.int:
+    return err LeopardError(code: LeopardNotEnoughData,
+      msg: LeopardNotEnoughDataMsg)
+
+  elif data.len > code.data.int:
+    return err LeopardError(code: LeopardTooMuchData,
+      msg: $leoResultString(wrapper.LeopardTooMuchData))
+
+  if symbolBytes < MinBufferSize.int or symbolBytes mod MinBufferSize.int != 0:
+    return err LeopardError(code: LeopardInvalidSize,
+      msg: $leoResultString(wrapper.LeopardInvalidSize))
+
+  var
+    enData = newSeq[pointer](code.data)
+
+  for i in 0..<code.data:
+    if data[i].len != symbolBytes:
+      for i in 0..<code.data: freeAligned enData[i]
+      return err LeopardError(code: LeopardInconsistentSize,
+        msg: LeopardInconsistentSizeMsg)
+
+    enData[i] = allocAligned(symbolBytes, LEO_ALIGN_BYTES)
```

Comment:
moving memory is probably not what you'd want to do here, specially if `data` is passed by var (which it should).

Reply:
For [`moveMem`](https://nim-lang.org/docs/system.html#moveMem%2Cpointer%2Cpointer%2CNatural) the Nim manual has:
> Copies the contents from the memory at `source` to the memory at `dest`.
> Exactly size bytes will be copied.

The wording for [`copyMem`](https://nim-lang.org/docs/system.html#copyMem%2Cpointer%2Cpointer%2CNatural) is the same. The difference between them is that `moveMem` handles overlapping regions while `copyMem` cannot.

I went with `moveMem` since I wasn't sure, but `copyMem` seems a better choice if we're not dealing with overlapping regions and perhaps we're not.

But as to whether copying is needed if `data` is passed by var, does it depend on the type of var? If we have `data: var seq[seq[byte]]` don't we still need to copy the data into an aligned allocation as expected by Leopard-RS?

Reply:
> I would rather this be a thin wrapper around leopard and the hire level functionality is moved over to the dagger repo.

Fair enough.

I included both the lower level functionality (in `leopard/wrapper.nim`) and higher-level functionality (in `leopard.nim) in the same project since it allowed me to develop and test them together.

What do you see as the advantages of splitting them apart?

Reply:
> Fair enough.

My reasoning is twofold, on the one hand a thin wrapper allows us to stay as close to the original implementation as possible without having to invent new APIs, and thus reuse the already existing documentation, etc... On the other hand, integrating this into dagger (or another project), will most likely require more tweaking and abstractions/indirection on top of it. In other words, a simpler more flexible low level API is easier to extend since it makes fewer assumptions.

Reply:
> Yes, the pointers should be aligned and for that reason we might have to do a copy if it isn't, but that isn't type related as such.

It looks like we can use the `align` pragma for our buffer of `seq[seq[byte]]` - https://nim-lang.org/docs/manual.html#implementation-specific-pragmas-align-pragma and I would expect that to internally align the data, then we can simply pass its pointer it __should__ work.

Reply:
> It looks like we can use the `align` pragma for our buffer of `seq[seq[byte]]` - https://nim-lang.org/docs/manual.html#implementation-specific-pragmas-align-pragma and I would expect that to internally align the data, then we can simply pass its pointer it **should** work.

I'll give it a shot.

Reply:
I think align is used to make sure an object is padded to a certain size, since we use byte it shouldn't have any impact